cpp/src/io/text/multibyte_split.cu

```diff
@@ -247,18 +307,137 @@ std::vector<rmm::cuda_stream_view> get_streams(int32_t count, rmm::cuda_stream_p
+
+  output_chunks(size_type max_output_size,
+                rmm::cuda_stream_view stream,
+                rmm::mr::device_memory_resource* mr)
+    : _size{0}, _max_output_size{max_output_size}
+  {
+    assert(max_output_size > 0);
+    _chunks.emplace_back(0, stream, mr);
+    _chunks.back().reserve(max_output_size * 2, stream);
+  }
+
+  output_chunks(output_chunks&&)      = delete;
+  output_chunks(const output_chunks&) = delete;
+  output_chunks& operator=(output_chunks&&) = delete;
+  output_chunks& operator=(const output_chunks&) = delete;
+
+  [[nodiscard]] split_device_span<T> next_output(rmm::cuda_stream_view stream)
+  {
+    auto head_it   = _chunks.end() - (_chunks.size() > 1 && _chunks.back().is_empty() ? 2 : 1);
+    auto head_span = free_span(*head_it);
+    if (head_span.size() >= _max_output_size) { return split_device_span<T>{head_span}; }
+    if (head_it == _chunks.end() - 1) {
+      // insert a new vector of double size
+      auto const next_chunk_size = 2 * _chunks.back().capacity();
+      _chunks.emplace_back(0, stream, _chunks.back().memory_resource());
+      _chunks.back().reserve(next_chunk_size, stream);
+    }
+    auto tail_span = free_span(_chunks.back());
+    assert(head_span.size() + tail_span.size() >= _max_output_size);
+    return split_device_span<T>{head_span, tail_span};
+  }
+
+  void advance_output(size_type actual_size)
+  {
+    assert(actual_size <= _max_output_size);
+    // this dummy stream is used for resizing, since we know we won't reallocate,
+    // providing a the correct stream is not necessary.
+    rmm::cuda_stream_view dummy_stream{};
+    if (_chunks.size() < 2) {
+      auto const new_size = _chunks.back().size() + actual_size;
+      assert(new_size <= _chunks.back().capacity());
+      _chunks.back().resize(new_size, dummy_stream);
+    } else {
+      auto& tail              = _chunks.back();
+      auto& prev              = *(_chunks.end() - 2);
+      auto const prev_advance = std::min(actual_size, prev.capacity() - prev.size());
+      auto const tail_advance = actual_size - prev_advance;
+      assert(tail.size() + tail_advance <= tail.capacity());
+      prev.resize(prev.size() + prev_advance, dummy_stream);
+      tail.resize(tail.size() + tail_advance, dummy_stream);
+    }
+    _size += actual_size;
+  }
+
+  [[nodiscard]] const rmm::device_uvector<T>& first_chunk() const { return _chunks.front(); }
+
+  [[nodiscard]] const rmm::device_uvector<T>& last_nonempty_chunk() const
+  {
+    return _chunks.size() > 1 && _chunks.back().is_empty() ? *(_chunks.end() - 2) : _chunks.back();
+  }
+
+  [[nodiscard]] size_type size() const
+  {
+    assert(_size ==
+           std::accumulate(
+             _chunks.begin(), _chunks.end(), size_type{}, [](size_type acc, const auto& chunk) {
+               return acc + chunk.size();
+             }));
+    return _size;
+  }
+
+  rmm::device_uvector<T> collect(rmm::cuda_stream_view stream,
```

Comment:
Do you think having an empty "moved-from"/"released-from" state makes sense here? I wanted to rely on the chunk vector never being empty, but maybe that's unnecessary?

Reply:
> I wanted to rely on the chunk vector never being empty, but maybe that's unnecessary?
if you "release" the underlying resources, I'd just make sure the struct is in the same state it would be in after the initial construction. That said, I don't think we need to worry too much about this class being reusable without doing a bigger design review and relocating it to a more accessible place.